### PR TITLE
py-dipy: Update version to support python@3.10

### LINKED ISSUE #40228: Installation issue: py-dipy@1.4.1 ^python@3.10
### Steps to reproduce the issue

```console
$ spack spec -II py-dipy
Input spec
--------------------------------
 -   py-dipy

Concretized
--------------------------------
 -   maobg6o  py-dipy@1.4.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  pkcpzcq      ^py-cython@0.29.33%gcc@11.3.0 build_system=python_pip patches=71de066 arch=linux-rhel8-zen2
 -   mg4eg4n      ^py-h5py@3.8.0%gcc@11.3.0~mpi build_system=python_pip arch=linux-rhel8-zen2
[+]  z244r5d          ^hdf5@1.14.0%gcc@11.3.0+cxx+fortran+hl+ipo~java+map~mpi+shared+szip+threadsafe+tools api=default build_system=cmake build_type=RelWithDebInfo generator=make patches=0b5dd6f arch=linux-rhel8-zen2
[+]  simblvj              ^cmake@3.26.3%gcc@11.3.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-rhel8-zen2
[+]  bcadp5n              ^gmake@4.4.1%gcc@11.3.0+guile build_system=autotools arch=linux-rhel8-zen2
[+]  fny73jv                  ^guile@2.2.6%gcc@11.3.0+readline build_system=autotools threads=none arch=linux-rhel8-zen2
[+]  zy5n3jc                      ^bdw-gc@8.2.2%gcc@11.3.0+libatomic-ops build_system=autotools threads=none arch=linux-rhel8-zen2
[+]  tkxbqtj                          ^libatomic-ops@7.8.0%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  rbqwy2o                      ^gmp@6.2.1%gcc@11.3.0+cxx build_system=autotools libs=shared,static patches=69ad2e2 arch=linux-rhel8-zen2
[+]  zyvinkw              ^libszip@2.1.1%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  qp6lvuk          ^py-pkgconfig@1.5.5%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  f4ovreq              ^py-poetry-core@1.2.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  pwyaspy      ^py-nibabel@4.0.2%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  zfccli6      ^py-numpy@1.24.3%gcc@11.3.0+blas+lapack build_system=python_pip patches=873745d arch=linux-rhel8-zen2
[+]  prjehhd          ^openblas@0.3.23%gcc@11.3.0~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=openmp arch=linux-rhel8-zen2
[+]  sve5wjh              ^perl@5.36.0%gcc@11.3.0+cpanm+open+shared+threads build_system=generic arch=linux-rhel8-zen2
[+]  nayr6pc                  ^berkeley-db@18.1.40%gcc@11.3.0+cxx+docs+stl build_system=autotools patches=26090f4 arch=linux-rhel8-zen2
[+]  qaqs5wt      ^py-packaging@23.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  zekphjz          ^py-flit-core@3.7.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  3juduhn      ^py-pip@23.0%gcc@11.3.0 build_system=generic arch=linux-rhel8-zen2
[+]  gzvj6wb      ^py-scipy@1.10.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  wqtnva5          ^pkg-config@0.29.2%gcc@11.3.0+internal_glib build_system=autotools arch=linux-rhel8-zen2
[+]  zluj75m          ^py-build@0.10.0%gcc@11.3.0+virtualenv build_system=python_pip arch=linux-rhel8-zen2
[+]  pcrgmx3              ^py-pyproject-hooks@1.0.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  6zh7lyq              ^py-tomli@2.0.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  3a6gd7i              ^py-virtualenv@20.22.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  2phfpyv                  ^py-distlib@0.3.6%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  xbjbqce                  ^py-filelock@3.12.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  ta44cdn                  ^py-hatch-vcs@0.3.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  drdbngf                  ^py-hatchling@1.14.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  fhkk7y5                      ^py-editables@0.3%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  lru6vlk                      ^py-pathspec@0.11.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  pj4qcf5                      ^py-pluggy@1.0.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  i67c6wa                      ^py-trove-classifiers@2023.3.9%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  nzcuxwn                          ^py-calver@2022.6.26%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  y6gzrwv                  ^py-platformdirs@3.5.0%gcc@11.3.0~wheel build_system=python_pip arch=linux-rhel8-zen2
[+]  nxwjenx          ^py-meson-python@0.12.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  na74xbv              ^meson@1.1.0%gcc@11.3.0 build_system=python_pip patches=0f0b1bd arch=linux-rhel8-zen2
[+]  6h5qprn              ^py-pyproject-metadata@0.7.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  oyn7aw2          ^py-pybind11@2.10.1%gcc@11.3.0+ipo build_system=cmake build_type=RelWithDebInfo generator=ninja arch=linux-rhel8-zen2
[+]  3gyusxe              ^ninja@1.11.1%gcc@11.3.0+re2c build_system=generic arch=linux-rhel8-zen2
[+]  35p2wk5                  ^re2c@2.2%gcc@11.3.0 build_system=generic arch=linux-rhel8-zen2
[+]  aue5kvd          ^py-pythran@0.12.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  g2uknuj              ^py-beniget@0.4.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  ws34kae              ^py-gast@0.5.3%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  rjq34l2              ^py-ply@3.11%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  63t3joz      ^py-setuptools@63.4.3%gcc@11.3.0 build_system=generic arch=linux-rhel8-zen2
[+]  s6ptrd4      ^py-tqdm@4.65.0%gcc@11.3.0+notebook+telegram build_system=python_pip arch=linux-rhel8-zen2
[+]  2wyvkwc          ^py-ipywidgets@8.0.2%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  avwaeov              ^py-ipykernel@6.22.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  q7nr2av                  ^py-comm@0.1.3%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  mxsfz6u                  ^py-debugpy@1.6.7%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  qe6dcly                  ^py-jupyter-client@7.1.2%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  4jwhxab                      ^py-entrypoints@0.4%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  4bg3tt2                      ^py-python-dateutil@2.8.2%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  p2p5tlc                  ^py-jupyter-core@5.3.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  yx4x67m                  ^py-matplotlib-inline@0.1.6%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  22ypx53                  ^py-nest-asyncio@1.5.6%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  idsjkbc                  ^py-psutil@5.9.4%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  tdmhq2o                  ^py-pyzmq@25.0.2%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  yzlebbp                      ^libzmq@4.3.4%gcc@11.3.0~docs+drafts+libbsd+libsodium+libunwind build_system=autotools patches=310b8aa arch=linux-rhel8-zen2
[+]  qwd5dcd                          ^libsodium@1.0.18%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  4iuwrq7                          ^libunwind@1.6.2%gcc@11.3.0~block_signals~conservative_checks~cxx_exceptions~debug~debug_frame+docs+pic+tests+weak_backtrace+xz+zlib build_system=autotools components=coredump,ptrace,setjump libs=shared,static arch=linux-rhel8-zen2
[+]  ey2jciq                      ^py-gevent@1.5.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  bsxhvms                          ^py-greenlet@2.0.2%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  kicr6yx                  ^py-tornado@6.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  55jc654              ^py-ipython@8.11.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  ux7ovr3                  ^py-backcall@0.2.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  rv7o22i                  ^py-decorator@5.1.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  6suj22u                  ^py-jedi@0.18.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  g7wzerq                      ^py-parso@0.8.3%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  wz4rx4e                  ^py-pexpect@4.8.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  oznzhuu                      ^py-ptyprocess@0.7.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  uufmszw                  ^py-pickleshare@0.7.5%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  kkfwruy                  ^py-prompt-toolkit@3.0.31%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  p5pcpj4                      ^py-wcwidth@0.2.5%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  nxu7fa2                  ^py-pygments@2.13.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  q33zdi4                  ^py-stack-data@0.5.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  qcpimia                      ^py-asttokens@2.2.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  f2mjwjb                      ^py-executing@1.1.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  vqbfgdj                      ^py-pure-eval@0.2.2%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  f7egrzd              ^py-jupyterlab-widgets@3.0.3%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  b4752ao              ^py-traitlets@5.9.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  js5udrj              ^py-widgetsnbextension@4.0.3%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  ostgo6z                  ^py-jupyter-packaging11@0.12.3%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  7pozavo                      ^py-deprecation@2.1.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  vbpvze2                      ^py-tomlkit@0.11.4%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  qrjceyy          ^py-requests@2.28.2%gcc@11.3.0+socks build_system=python_pip arch=linux-rhel8-zen2
[+]  q7uerfy              ^py-certifi@2023.5.7%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  ehzzsfw              ^py-charset-normalizer@2.0.12%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  6fe7ziw              ^py-idna@3.4%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  r5hmcig              ^py-pysocks@1.7.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  jxmzzhn              ^py-urllib3@1.26.12%gcc@11.3.0+brotli+secure+socks build_system=python_pip arch=linux-rhel8-zen2
[+]  rcydzcf                  ^py-brotlipy@0.7.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  rahg5ar                      ^py-cffi@1.15.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  5mxdt5f                          ^py-pycparser@2.21%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  ihuciyv                  ^py-cryptography@40.0.2%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  v6jtsxw                      ^git@2.40.0%gcc@11.3.0+man+nls+perl+subtree~svn~tcltk build_system=autotools arch=linux-rhel8-zen2
[+]  ml7o4bu                          ^curl@7.88.1%gcc@11.3.0+gssapi~ldap+libidn2~librtmp+libssh+libssh2+nghttp2 build_system=autotools libs=shared,static tls=gnutls,mbedtls,openssl arch=linux-rhel8-zen2
[+]  vvo2utu                              ^gnutls@3.7.8%gcc@11.3.0+guile+zlib build_system=autotools arch=linux-rhel8-zen2
[+]  zksb6q3                                  ^nettle@3.8.1%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  5et2bgd                              ^libssh@0.8.9%gcc@11.3.0+gssapi+ipo build_system=cmake build_type=RelWithDebInfo generator=make arch=linux-rhel8-zen2
[+]  ek4c3lz                              ^mbedtls@2.28.1%gcc@11.3.0+pic build_system=makefile build_type=RelWithDebInfo libs=shared,static arch=linux-rhel8-zen2
[+]  ddcqjdi                              ^nghttp2@1.52.0%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  z4v5kwp                          ^libidn2@2.3.4%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  r4hghfg                          ^openssh@9.3p1%gcc@11.3.0+gssapi build_system=autotools arch=linux-rhel8-zen2
[+]  egb5eml                              ^libedit@3.1-20210216%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  ywbf4to                          ^pcre2@10.42%gcc@11.3.0+jit+multibyte build_system=autotools arch=linux-rhel8-zen2
[+]  aonl62u                      ^py-setuptools-rust@1.5.1%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  7l2dkp7                          ^py-semantic-version@2.10.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  z3zmdxv                      ^rust@1.65.0%gcc@11.3.0+analysis+clippy~rls+rustfmt+src build_system=generic extra_targets=none arch=linux-rhel8-zen2
[+]  lplm56x                          ^libgit2@1.6.4%gcc@11.3.0~curl+ipo+mmap+ssh build_system=cmake build_type=RelWithDebInfo generator=make https=system arch=linux-rhel8-zen2
[+]  hlnp7va                              ^pcre@8.45%gcc@11.3.0+jit+multibyte+utf build_system=autotools arch=linux-rhel8-zen2
[+]  glfk6ep                          ^libssh2@1.10.0%gcc@11.3.0+ipo+shared build_system=cmake build_type=RelWithDebInfo crypto=mbedtls generator=make arch=linux-rhel8-zen2
[+]  jcx42cc                  ^py-pyopenssl@19.0.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  gqmbuac                      ^py-six@1.16.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  y3m7i75                  ^py-urllib3-secure-extra@0.1.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  qhp4bdf          ^py-setuptools-scm@7.0.5%gcc@11.3.0+toml build_system=python_pip arch=linux-rhel8-zen2
[+]  fiekb6r              ^py-typing-extensions@4.5.0%gcc@11.3.0 build_system=python_pip arch=linux-rhel8-zen2
[+]  yiqallw      ^py-wheel@0.37.1%gcc@11.3.0 build_system=generic arch=linux-rhel8-zen2
[+]  qoumw5v      ^python@3.10.10%gcc@11.3.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma+nis+optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,7d40923,f2fd060 arch=linux-rhel8-zen2
[+]  y2p2gbz          ^bzip2@1.0.8%gcc@11.3.0~debug+pic+shared build_system=generic arch=linux-rhel8-zen2
[+]  nbfheoo              ^diffutils@3.9%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  edn5eua          ^expat@2.5.0%gcc@11.3.0+libbsd build_system=autotools arch=linux-rhel8-zen2
[+]  h5im2xa              ^libbsd@0.11.7%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  5sslvdw                  ^libmd@1.0.4%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  tspio6g          ^gdbm@1.23%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  aym4bs5          ^gettext@0.21.1%gcc@11.3.0+bzip2+curses+git+libunistring+libxml2+tar+xz build_system=autotools arch=linux-rhel8-zen2
[+]  ghqoznx              ^libiconv@1.17%gcc@11.3.0 build_system=autotools libs=shared,static arch=linux-rhel8-zen2
[+]  zra6tcb              ^libunistring@1.1%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  hfk7szf              ^libxml2@2.10.3%gcc@11.3.0~python build_system=autotools arch=linux-rhel8-zen2
[+]  vjlzfma              ^tar@1.34%gcc@11.3.0 build_system=autotools zip=pigz arch=linux-rhel8-zen2
[+]  pnsotvl                  ^pigz@2.7%gcc@11.3.0 build_system=makefile arch=linux-rhel8-zen2
[+]  5hgslof                  ^zstd@1.5.5%gcc@11.3.0+programs build_system=makefile compression=lz4,lzma,zlib libs=shared,static arch=linux-rhel8-zen2
[+]  qmkbina                      ^lz4@1.9.4%gcc@11.3.0 build_system=makefile libs=shared,static arch=linux-rhel8-zen2
[+]  akxkimj          ^libffi@3.4.4%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  5wi2hwp          ^libnsl@2.0.0%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  yxpydi6              ^autoconf@2.69%gcc@11.3.0 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-rhel8-zen2
[+]  vr5d2na              ^automake@1.16.5%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  x3keyoo              ^libtirpc@1.2.6%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  hoxnvw7                  ^krb5@1.20.1%gcc@11.3.0+shared build_system=autotools arch=linux-rhel8-zen2
[+]  idcnjva                      ^bison@3.8.2%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  ovq4gsp              ^libtool@2.4.7%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  2xgh4yr              ^m4@1.4.19%gcc@11.3.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-rhel8-zen2
[+]  lskfqtb                  ^libsigsegv@2.14%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  bgjnz6u              ^rpcsvc-proto@1.4.3%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  jeqzbqg          ^libuuid@1.0.3%gcc@11.3.0 build_system=autotools arch=linux-rhel8-zen2
[+]  5zyvitn          ^libxcrypt@4.4.33%gcc@11.3.0~obsolete_api build_system=autotools arch=linux-rhel8-zen2
[+]  k6m5lie          ^ncurses@6.4%gcc@11.3.0+symlinks+termlib abi=5 build_system=autotools arch=linux-rhel8-zen2
[+]  bjvjttn          ^openssl@1.1.1v%gcc@11.3.0~docs+shared build_system=generic certs=system arch=linux-rhel8-zen2
[+]  boheg7m          ^readline@8.2%gcc@11.3.0 build_system=autotools patches=bbf97f1 arch=linux-rhel8-zen2
[+]  5qlirxs          ^sqlite@3.40.1%gcc@11.3.0+column_metadata+dynamic_extensions+fts+functions+rtree build_system=autotools arch=linux-rhel8-zen2
[+]  ranwi3u          ^xz@5.4.1%gcc@11.3.0+pic build_system=autotools libs=shared,static arch=linux-rhel8-zen2
[+]  gv2fyi6          ^zlib@1.2.13%gcc@11.3.0+optimize+pic+shared build_system=makefile arch=linux-rhel8-zen2
...
```

### Error message

<details><summary>Error message</summary>

<pre>
...
  building 'dipy.reconst.recspeed' extension
  /software/spack/git.2023.05.30/lib/spack/env/gcc/gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -fPIC -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -Isrc -I/cvmfs/hpcsw.umd.edu/spack-software/2023.05.30/linux-rhel8-zen2/gcc-11.3.0/py-numpy-1.24.3-zfccli6u6rgt7q6zq7njmxl7epza5f6u/lib/python3.10/site-packages/numpy/core/include -Ibuild -I/cvmfs/hpcsw.umd.edu/spack-software/2023.05.30/linux-rhel8-zen2/gcc-11.3.0/python-3.10.10-qoumw5vkaizey7uxv4524pmdpcqdfqen/include/python3.10 -c dipy/reconst/recspeed.c -o build/temp.linux-x86_64-cpython-310/dipy/reconst/recspeed.o -msse2 -mfpmath=sse -fopenmp
  dipy/reconst/recspeed.c: In function '__Pyx_call_return_trace_func':
  dipy/reconst/recspeed.c:1881:15: error: 'PyThreadState' {aka 'struct _ts'} has no member named 'use_tracing'; did you mean 'tracing'?
   1881 |       tstate->use_tracing = 0;
        |               ^~~~~~~~~~~
        |               tracing
  dipy/reconst/recspeed.c:1887:15: error: 'PyThreadState' {aka 'struct _ts'} has no member named 'use_tracing'; did you mean 'tracing'?
   1887 |       tstate->use_tracing = 1;
        |               ^~~~~~~~~~~
        |               tracing
  dipy/reconst/recspeed.c: In function '__pyx_pf_4dipy_7reconst_8recspeed_4local_maxima':
  dipy/reconst/recspeed.c:1832:32: error: 'PyThreadState' {aka 'struct _ts'} has no member named 'use_tracing'; did you mean 'tracing'?
   1832 |           if (unlikely(tstate->use_tracing) && !tstate->tracing &&\
        |                                ^~~~~~~~~~~
  dipy/reconst/recspeed.c:811:43: note: in definition of macro 'unlikely'
    811 |   #define unlikely(x) __builtin_expect(!!(x), 0)
        |                                           ^
  dipy/reconst/recspeed.c:6977:3: note: in expansion of macro '__Pyx_TraceCall'
   6977 |   __Pyx_TraceCall("local_maxima", __pyx_f[0], 202, 0, __PYX_ERR(0, 202, __pyx_L1_error));
        |   ^~~~~~~~~~~~~~~
... [and similar errors]
</pre></details>

The error report is rather long, but the above is a smaple.  Full report can be found in attached spack-build-out.txt.
The above errors appear to be related to changes in python@3.10, and py-dipy does not support python@3.10 until py-dipy@1.5.0.

### Information on your system

```console
 spack debug report
* **Spack:** 0.21.0.dev0 (84faf5a6cf342efddb976b7432e44aff12f1c985)
* **Python:** 3.6.8
* **Platform:** linux-rhel8-zen2
* **Concretizer:** clingo
```

### Additional information

Maintainers are @adamjstewart  @pradyunsg

[spack-build-env.txt](https://github.com/spack/spack/files/12742201/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/12742202/spack-build-out.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by payerle**
py-dipy only adds support to python@3.10 in py-dipy@1.5.0

Closes #40228

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-dipy/package.py [hunk 1]
```diff
@@ -19,9 +19,11 @@ class PyDipy(PythonPackage):
     homepage = "https://dipy.org/"
     pypi = "dipy/dipy-1.4.1.tar.gz"
 
+    version("1.7.0", sha256="59bb647128aae7793215c813bb8ea35dae260ac9f0d938c724064f0af5a05cc3")
     version("1.4.1", sha256="b4bf830feae751f3f985d54cb71031fc35cea612838320f1f74246692b8a3cc0")
 
     depends_on("python@3.6:", type=("build", "run"))
+    depends_on("python@:3.9", type=("build", "run"), when="@:1.4")
```
Comment (manuelakuhn) [inline var/spack/repos/builtin/packages/py-dipy/package.py:26]: optional:
switch the order of the `type` and `when` argument as most of the other python packages do

**Review by manuelakuhn — COMMENTED** [review]
The minimum version requirements changed for 1.7.0. Can you add them?

CYTHON_MIN_VERSION = '0.29.24'
NUMPY_MIN_VERSION = '1.14.5'
SCIPY_MIN_VERSION = '1.1'
NIBABEL_MIN_VERSION = '3.0.0'
H5PY_MIN_VERSION = '2.8.0'
(see [dipy/info.py](https://github.com/dipy/dipy/blob/1.7.0/dipy/info.py#L77C1-L81C27))

**Review by manuelakuhn — APPROVED** [review]
If you want to be exact, the new restricted dependencies are present starting from `py-dipy@1.5:`
```suggestion
    depends_on("py-cython@0.29.24:", type=("build"), when="@1.5:")
    depends_on("py-numpy@1.14.5:", type=("build", "run"), when="@1.5:")
    depends_on("py-scipy@1.1:", type=("build", "run"), when="@1.5:")
    depends_on("py-h5py@2.8.0:", type=("build", "run"), when="@1.5:")
```

**Review by adamjstewart — APPROVED** [review]
Good enough

**Pull request merged**